### ISSUE TITLE
Response protocol not efficient enough for socket programming

### ISSUE THREAD

**Issue by GoogleCodeExporter** [ashish21590/redis#243]
```
Redis is a fast database. I am using Python to connect to Redis. I tested 
redis-py but was not 
impressed with the speed. The main issue seems to be efficient socket 
programming in Python 
and not the Redis database. So, I tried to write my own Python Redis client. 
Please note that I am 
just a medior Python programmer.

This article (http://www.amk.ca/python/howto/sockets/) has the following to say 
about sockets: 

"I repeat: if a socket send or recv returns after handling 0 bytes, the 
connection has been broken. 
If the connection has not been broken, you may wait on a recv forever, because 
the socket will 
not tell you that there's nothing more to read (for now). Now if you think 
about that a bit, you'll 
come to realize a fundamental truth of sockets: messages must either be fixed 
length (yuck), or 
be delimited (shrug), or indicate how long they are (much better), or end by 
shutting down the 
connection. The choice is entirely yours, (but some ways are righter than 
others)."

Redis seems to be using a delimited response. This means I have to do a lot of 
socket.recv's to 
retrieve a response.

For example. In a multi-bulk response you can find out the number of items that 
is going to be 
returned. But, there is no information on how many bytes the total response is 
going to be. So, 
you are left with the only option of doing multiple socket.recv's to retrieve 
the response.

So, I have the following feature request: include the number of bytes of the 
TOTAL response at 
the beginning of each response (especially multi-bulk responses). This gives 
client programmers 
the opportunity to retrieve the response in one call and then proces the 
response on client side. 
This could improve the performance of pipelined requests even further.

For example, change this:

C: LRANGE mylist 0 3
S: *4
S: $3
S: foo
S: $3
S: bar
S: $5
S: Hello
S: $5
S: World

into this:

C: LRANGE mylist 0 3
S: *4$24
S: $3
S: foo
S: $3
S: bar
S: $5
S: Hello
S: $5
S: World

*4$24 indicates that 4 items are going to be send with a total length of 24 
bytes (excluding the 
\r\n's - in this case 4 * 2 * 2 bytes = 16 bytes).

Thanks.
```

Original issue reported on code.google.com by `berry.gr...@gmail.com` on 13 May 2010 at 2:47

**Comment by GoogleCodeExporter** [comment]
```
Why don't you read as much as you can into a local buffer and then parse that?
```

Original comment by `mel...@gmail.com` on 13 May 2010 at 3:04

**Comment by GoogleCodeExporter** [comment]
```
Hello Berry,

probably what you are noticing is the delay due to the round trip time, not 
something related to parsing 
speed.
The problem with speed can be fixed with a buffer if you really need to go 
faster, but I bet this is not going to 
be your bottleneck.

The are several reasons why there is not a prefixed length:
- In our benchmarks this is not the real problem.
- when replies are big, to read the whole thing is not a good idea anyway. the 
current format is designed to be 
parsed as a stream.
- to buffer the whole reply in order to be able to attach a prefix length is 
slow or tricky depending on the 
implementation... there is the need to place sentinels in the output buffer and 
so forth.
- We'll add UDP for very high performance read queries

Cheers,
Salvatore
```

Original comment by `anti...@gmail.com` on 13 May 2010 at 3:15

**Comment by GoogleCodeExporter** [comment]
```
Using buffered reads drastically reduces the number of socket.recv's and thus 
improved the performance. Thanks 
for the tip.
```

Original comment by `berry.gr...@gmail.com` on 13 May 2010 at 10:39

**Comment by GoogleCodeExporter** [comment]
```
Please feel free to contribute patches to redis-py if you find ways to speed 
things up.
```

Original comment by `sed...@gmail.com` on 14 May 2010 at 12:55

**Comment by GoogleCodeExporter** [comment]
```
@sedrik I am still investigating. If I find some useful improvements I will 
contribute patches to redis-py.
```

Original comment by `berry.gr...@gmail.com` on 14 May 2010 at 5:30

**Comment by GoogleCodeExporter** [comment]
```
I have to agree with the OP. I've built a bson parser (for the C# MongoDB 
drivers),
and such a protocol is easier to parse, faster to parse, and causes less memory
fragmentation. Sure, its relatively trivial if you wanna use a blocking 
operation and
treat the TCP stream like a filestream and execute a ReadLine(), but if you 
wanna use
asynchronous callbacks and enable connection pooling - data fragmentation is a 
pain
to deal with. Even something as basic as having a 2-byte delimiter causes 
unnecessary
complication.

The docs give the following as a sample reply to GET mykey:
$6\r\nfoobar\r\n

I think this would be much simpler:
$6foobar

where 6 is a 4-byte integer. This wouldn't only make the data stream slightly 
smaller
but would also make parsing much easier.

```

Original comment by `google84...@openmymind.io` on 19 May 2010 at 12:48

**Comment by GoogleCodeExporter** [comment]
```
I don't see why BSON would be fundamentally easier or faster to parse. The only 
difference between BSON and 
this protocol is the newlines, and for a good reason. The newlines are in the 
protocol so debugging a stream of 
data is as simple as attaching a netcat, and get readable data, not for 
allowing people to use readline(). Moreover, 
I would highly suggest against this. Every argument is prefixed with its 
length, so there is no need to use 
readline(). Even better, every client lib author can simply do a read() and 
start traversing the buffer, using a 
simple index. Not that complex nor slow in my opinion.
```

Original comment by `pcnoordh...@gmail.com` on 19 May 2010 at 8:25

**Comment by GoogleCodeExporter** [comment]
```
Some actual benchmarks would be immensely useful in this discussion.  Can one 
of the 
proponents of the change perform some?
```

Original comment by `jzaw...@gmail.com` on 19 May 2010 at 2:04

**Comment by GoogleCodeExporter** [comment]
```
@jzawodn: I don't have benchmarks but I've some other info that is strongly 
correlated, that is, apparently the 
profiling efforts on the Redis code does not show the parsing code as a big 
culprit for the CPU usage. There 
are tons of things to optimize before of the protocol. For instance the command 
lookup is totally lame, 
designed for a server with < 10 commands, as it's an O(N) function not even 
using binary search. I already 
know how to fix many of this issues, just not possible around RC1 time.

Many of this optimizations are planned for 2.2, that I hope will be a 
just-go-faster-smaller release. Then I'll 
profile again, but current data is indicating that the speed gain can be very 
little, so currently there are no 
plans to switch to another protocol *for TCP*.

Instead the UDP protocol will surely be binary, and I've already a draft of the 
protocol.

Cheers,
Salvatore
```

Original comment by `anti...@gmail.com` on 19 May 2010 at 2:50

**Comment by GoogleCodeExporter** [comment]
Original comment by `anti...@gmail.com` on 31 Aug 2010 at 11:03
- Changed state: **WontFix**

**Comment by GoogleCodeExporter** [comment]
```
I have been doing some benchmarking recently, and am dismayed by the python 
redis  module speed. For a lot of my use cases (10-100k lrange queries) python 
redis is roughly 6x slower than phpredis. That's a serious bummer! I also have 
been thinking it's socket-related. Is there any news on this, or are those 
patches floating around?
```

Original comment by `pet...@myprofitsee.com` on 23 Feb 2011 at 9:02

**Comment by GoogleCodeExporter** [comment]
```
Andy merged patches a couple of weeks ago that add in support for hiredis-py. 
This module is a binding to Redis's own hiredis C client, especially created 
for speeding up the kind of queries you mention (large multi bulks). You can 
check it out here: https://github.com/pietern/hiredis-py

When you install it, the latest version of redis-py (maybe you need to use the 
code straight from the repository) should be able to pick it up automatically.
```

Original comment by `pcnoordh...@gmail.com` on 23 Feb 2011 at 9:07

**Comment by GoogleCodeExporter** [comment]
```
I saw some pyhthon hiredis wrapper [1] build by Pieter Noordhuis, and I heard 
something about improving performance 

[1] http://pypi.python.org/pypi/hiredis/0.1.0
```

Original comment by `pfrei...@gmail.com` on 23 Feb 2011 at 9:07

**Comment by GoogleCodeExporter** [comment]
```
I've used pieter's module lately and it rocks. for big bulk responses it kicks 
ass, can be almost x100 faster. 

I am also now actively (but slowly :) ) working on a pure C python extension, 
also based on hiredis internally, but without any python code wrapping it. It 
works and it's very fast, especially for small to medium responses (where 
pieter's module is not much faster than redis-py, since not much parsing is 
going on), but it's far from being complete in terms of the number of commands 
it covers. If anyone's willing to help along implementing more callls, I'd be 
glad to share it.
```

Original comment by `dvir...@gmail.com` on 23 Feb 2011 at 9:54
